packages/gatsby-plugin-gatsby-cloud/src/build-headers-program.js

```diff
@@ -0,0 +1,348 @@
+    )
+    hasPageData = existsSync(pageDataPath)
+  }
+
+  pages.forEach(page => {
+    const scripts = _.flatMap(COMMON_BUNDLES, file =>
+      getScriptPath(file, manifest)
+    )
+    scripts.push(...getScriptPath(pathChunkName(page.path), manifest))
+    scripts.push(...getScriptPath(page.componentChunkName, manifest))
+
+    const json = []
+    if (hasAppData) {
+      json.push(posix.join(PAGE_DATA_DIR, `app-data.json`))
+    }
+
+    if (hasPageData) {
+      json.push(getPageDataPath(page.path))
+    }
+
+    const filesByResourceType = {
+      script: scripts.filter(Boolean),
+      fetch: json,
+    }
+
+    const pathKey = headersPath(pathPrefix, page.path)
+    linksByPage[pathKey] = linkHeaders(filesByResourceType, pathPrefix)
+  })
+
+  return linksByPage
+}
+
+function defaultMerge(...headers) {
+  function unionMerge(objValue, srcValue) {
+    if (_.isArray(objValue)) {
+      return _.union(objValue, srcValue)
+    } else {
+      return undefined // opt into default merge behavior
+    }
+  }
+
+  return _.mergeWith({}, ...headers, unionMerge)
+}
+
+function headersMerge(userHeaders, defaultHeaders) {
+  const merged = {}
+  Object.keys(defaultHeaders).forEach(path => {
+    if (!userHeaders[path]) {
+      merged[path] = defaultHeaders[path]
+      return
+    }
+    const headersMap = {}
+    defaultHeaders[path].forEach(header => {
+      headersMap[getHeaderName(header)] = header
+    })
+    userHeaders[path].forEach(header => {
+      headersMap[getHeaderName(header)] = header // override if exists
+    })
+    merged[path] = Object.values(headersMap)
+  })
+  Object.keys(userHeaders).forEach(path => {
+    if (!merged[path]) {
+      merged[path] = userHeaders[path]
+    }
+  })
+  return merged
+}
+
+function transformLink(manifest, publicFolder, pathPrefix) {
+  return header =>
+    header.replace(LINK_REGEX, (__, prefix, file, suffix) => {
+      const hashed = manifest[file]
+      if (hashed) {
+        return `${prefix}${pathPrefix}${hashed}${suffix}`
+      } else if (existsSync(publicFolder(file))) {
+        return `${prefix}${pathPrefix}${file}${suffix}`
+      } else {
+        throw new Error(
+          `Could not find the file specified in the Link header \`${header}\`.` +
+            `The gatsby-plugin-gastby-cloud is looking for a matching file (with or without a ` +
+            `webpack hash). Check the public folder and your gatsby-config.js to ensure you are ` +
+            `pointing to a public file.`
+        )
+      }
+    })
+}
+
+function stringifyHeaders(headers) {
+  return _.reduce(
+    headers,
+    (text, headerList, path) => {
+      const headersString = _.reduce(
+        headerList,
+        (accum, header) => `${accum}  ${header}\n`,
+        ``
+      )
+      return `${text}${path}\n${headersString}`
+    },
+    ``
+  )
+}
+
+// program methods
+
+const validateUserOptions = (pluginOptions, reporter) => headers => {
```

Comment:
Rather than validating this here, could this be done in [plugin options schema](https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/configuring-usage-with-plugin-options/#how-to-validate-plugin-options)

Reply:
This is a good idea